packages/vm/src/imports.rs

```diff
@@ -113,12 +153,21 @@ pub fn do_humanize_address<A: Api, S: Storage, Q: Querier>(
 ) -> VmResult<u32> {
     let canonical = Binary(read_region(ctx, source_ptr, MAX_LENGTH_CANONICAL_ADDRESS)?);
 
-    // TODO: how to report API errors back to the contract?
-    let (human, gas_info) = api.human_address(&CanonicalAddr(canonical))?;
-    process_gas_info::<S, Q>(ctx, gas_info)?;
-
-    write_region(ctx, destination_ptr, human.as_str().as_bytes())?;
-    Ok(0)
+    match api.human_address(&CanonicalAddr(canonical)) {
+        Ok((human, gas_info)) => {
+            process_gas_info::<S, Q>(ctx, gas_info)?;
+            write_region(ctx, destination_ptr, human.as_str().as_bytes())?;
+            Ok(0)
+        }
+        Err((FfiError::UserErr { msg, .. }, gas_info)) => {
+            process_gas_info::<S, Q>(ctx, gas_info)?;
+            Ok(write_to_contract::<S, Q>(ctx, msg.as_bytes())?)
```

Comment:
Trying to see how the contract distinguished success from failure.

On success, the return value is 0 and the passed in pointer contains data (of the human address).
On UserErr, the return value is non-zero and points to a byte array with the utf-8 encoded error message, but destination_ptr remains empty.

Correct?

Reply:
Right, the return value is an error pointer.

```rust
                // Reads human address from source_ptr and writes canonicalized representation to destination_ptr.
                // A prepared and sufficiently large memory Region is expected at destination_ptr that points to pre-allocated memory.
                // Returns 0 on success. Returns a non-zero memory location to a Region containing an UTF-8 encoded error string for invalid inputs.
                // Ownership of both input and output pointer is not transferred to the host.
                "canonicalize_address" => Func::new(move |ctx: &mut Ctx, source_ptr: u32, destination_ptr: u32| -> VmResult<u32> {
                    do_canonicalize_address::<A, S, Q>(api, ctx, source_ptr, destination_ptr)
                }),
                // Reads canonical address from source_ptr and writes humanized representation to destination_ptr.
                // A prepared and sufficiently large memory Region is expected at destination_ptr that points to pre-allocated memory.
                // Returns 0 on success. Returns a non-zero memory location to a Region containing an UTF-8 encoded error string for invalid inputs.
                // Ownership of both input and output pointer is not transferred to the host.
                "humanize_address" => Func::new(move |ctx: &mut Ctx, source_ptr: u32, destination_ptr: u32| -> VmResult<u32> {
                    do_humanize_address::<A, S, Q>(api, ctx, source_ptr, destination_ptr)
                }),
```

This has been implemented before, so you don't see the error propagation here.